libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs

```diff
@@ -54,7 +57,21 @@ public async override Task ExecuteAsync(TestAdapter adapter, BotCallbackHandler
                 {
                     foreach (var assignment in Assignments)
                     {
-                        dc.State.SetValue(assignment.Property.Value, assignment.Value.Value);
+                        JToken value = null;
+                        var (val, valueError) = assignment.Value.TryGetValue(dc.State);
+                        if (valueError != null)
+                        {
+                            throw new Exception($"Expression evaluation resulted in an error. Expression: {assignment.Value.ToString()}. Error: {valueError}");
+                        }
+
+                        if (val != null)
+                        {
+                            value = JToken.FromObject(val).DeepClone();
+                        }
+
+                        value = value?.ReplaceJTokenRecursively(dc.State);
+                        var property = assignment.Property.GetValue(dc.State);
+                        dc.State.SetValue(property, value);
```

Comment:
What was wrong with just calling SetValue?

Reply:
I use the logic from Microsoft.SetProperties. ReplaceJTokenRecursively will handle object binding correctly and I updated the test case for that:
```
                {
                    "property": "user.object",
                    "value": {
                        "value": "${user.object.value} again"
                    }
                }
```

Reply:
I'm still missing why you can't just do dc.State.SetValue(property, value) why do you need to deep clone and replace jtoken?  We don't need to do that anywhere else to use SetValue.

---
In reply to: [505218331](https://github.com/microsoft/botbuilder-dotnet/pull/4720#discussion_r505218331) [](ancestors = 505218331)